src/lib/paginator/paginator.spec.ts

```diff
@@ -0,0 +1,287 @@
+      expect(paginator.currentPageIndex).toBe(10);
+      expect(paginator.canNavigateToNextPage()).toBe(false);
+
+      component.latestPageChangeEvent = null;
+      component.clickNextButton();
+
+      expect(component.latestPageChangeEvent).toBe(null);
+      expect(paginator.currentPageIndex).toBe(10);
+    });
+
+    it('should disable navigating to the previous page if at first page', () => {
+      expect(paginator.currentPageIndex).toBe(0);
+      expect(paginator.canNavigateToPreviousPage()).toBe(false);
+
+      component.latestPageChangeEvent = null;
+      component.clickPreviousButton();
+
+      expect(component.latestPageChangeEvent).toBe(null);
+      expect(paginator.currentPageIndex).toBe(0);
+    });
+  });
+
+  it('should fail if page length and page length options are not provided', () => {
+    expect(() => {
+      TestBed.createComponent(MdPaginatorWithoutPageLengthOrOptionsApp).detectChanges();
+    }).toThrowError('No page length options available for the MdPaginator');
+  });
+
+  it('should default the page length options to the page length if no options provided', () => {
+    const withoutOptionsAppFixture = TestBed.createComponent(MdPaginatorWithoutOptionsApp);
+    withoutOptionsAppFixture.detectChanges();
+
+    expect(withoutOptionsAppFixture.componentInstance.mdPaginator._displayedPageLengthOptions)
+        .toEqual([10]);
+  });
+
+  it('should default the page length to the first page length option if no length provided', () => {
+    const withoutLengthAppFixture = TestBed.createComponent(MdPaginatorWithoutPageLengthApp);
+    withoutLengthAppFixture.detectChanges();
+
+    expect(withoutLengthAppFixture.componentInstance.mdPaginator._pageLength).toBe(10);
+  });
+
+  it('should show a sorted list of page length options including the current page length', () => {
+    expect(paginator._displayedPageLengthOptions).toEqual([5, 10, 25, 100]);
+
+    component.pageLength = 30;
+    fixture.detectChanges();
+    expect(paginator.pageLengthOptions).toEqual([5, 10, 25, 100]);
+    expect(paginator._displayedPageLengthOptions).toEqual([5, 10, 25, 30, 100]);
+
+    component.pageLengthOptions = [100, 25, 10, 5];
+    fixture.detectChanges();
+    expect(paginator._displayedPageLengthOptions).toEqual([5, 10, 25, 30, 100]);
+  });
+
+  it('should be able to change the page length while keeping the first item present', () => {
+    // Start on the third page of a list of 100 with a page length of 10.
+    component.currentPageIndex = 4;
+    component.pageLength = 10;
+    component.listLength = 100;
+    fixture.detectChanges();
+
+    // The first item of the page should be item with index 40
+    let firstPageItemIndex = paginator.currentPageIndex * paginator.pageLength;
+    expect(firstPageItemIndex).toBe(40);
+
+    // The first item on the page is now 25. Change the page length to 25 so that we should now be
+    // on the second page where the top item is index 25.
+    paginator._changePageLength(25);
```

Comment:
This seems like it's a private method. It might be better to test this through the `@Input`.